doc/TclList.3
'\"
'\" Copyright (c) 1996-1998 Sun Microsystems, Inc.
'\"
'\" See the file "license.terms" for information on usage and redistribution
'\" of this file, and for a DISCLAIMER OF ALL WARRANTIES.
'\" 
'\" RCS: @(#) $Id$
'\" 
.so man.macros
.TH TclList 3 1.0 Jacl "Jacl Library Procedures"
.BS
.SH CLASS
tcl.lang.TclList -- This class is used to create, modify and read from Tcl list
objects. 

.SH EXTENDS
tcl.lang.InternalRep

.SH METHODS
.sp
static TclObject \fBnewInstance\fR()
.sp
static void \fBappend\fR(Interp \fIinterp\fR, TclObject \fItobj\fR, TclObject \fIelem\fR)
throws TclException
.sp
static int \fBgetLength\fR(Interp \fIinterp\fR, TclObject \fItobj\fR)
throws TclException
.sp
static TclObject[ ] \fBgetElements\fR(Interp \fIinterp\fR, TclObject \fItobj\fR)
throws TclException
.sp
static TclObject \fBindex\fR(Interp \fIinterp\fR, TclObject \fItobj\fR, int \fIindex\fR)
throws TclException
.sp
static void \fBreplace\fR(Interp \fIinterp\fR, TclObject \fItobj\fR, int \fIindex\fR, int \fIcount\fR, TclObject \fIelts[]\fR, int \fIfrom\fR, int \fIto\fR)

.SH ARGUMENTS
.AS TclObject "elts[]" .
.AP Interp interp
If an error occurs while converting an object to be a list object,
an error message is left in the interpreter's result object
unless \fIinterp\fR is NULL.
.AP TclObject tobj
Points to the list object to be manipulated.
If \fItobj\fR does not already point to a list object,
an attempt will be made to convert it to one.
.AP TclObject elem
Points to a list object containing elements to be appended onto \fItobj\fR.
.AP TclObject elts[]
An array of TclObjects to reaplace those in the list object \fItobj\fR.
.AP int index
The index of the elemtent of \fItobj\fR in question.
.AP int count
The number of elemtents of the \fItobj\fR to replace.
.AP int from
The index of the first replacement element in \fIelts\fR.
.AP int to
The index of the last replacement element in \fIelts\fR.
.BE

.SH DESCRIPTION
.PP
The TclList class is used to create, modify and read from Tcl list
objects. All of the methods defined below attempt to change the
internal representation of \fItobj\fR parameters to TclList. If such a
conversion fails (e.g., the string value of \fItobj\fR is not a proper Tcl
list), a TclException is thrown, and an appropriate error message
is left in \fIinterp\fR's result if \fIinterp\fR is not null.
.PP
Tcl list objects have an internal representation that supports
the efficient indexing and appending.
The procedures described in this man page are used to
create, modify, index, and append to Tcl list objects from Java code.
In all cases where \fItobj\fR is an argument, if \fItobj\fR is not a
TclList, then an attempt is made to convert it to a TclList.  If
the attempt fails, a TclException is thrown.
.TP
\fBnewInstance\fR
The \fBnewInstance\fR method creates and returns a new empty TclList.
.TP
\fBappend\fR
The \fBappend\fR method adds one object, \fIelem\fR, to the end of the list
object referenced by \fItobj\fR.  After appending the element,
\fBappend\fR increments the element's reference count since \fItobj\fR
now also refers to it.
.TP
\fBgetLength\fR
The \fBgetLength\fR method returns the number of elements in \fItobj\fR.
.TP
\fBgetElements\fR
The \fBgetElements\fR method converts the TclList \fItobj\fR to an array of
TclObjects.  The objects referenced by the returned array should be
treated as readonly, and their ref counts are not incremented.
.TP
\fBindex\fR
The \fBindex\fR method returns a pointer to the \fIindex\fRth element of
\fItobj\fR.  If \fIindex\fR is negative or greater than or equal to
the number of elements in the list, null is returned.
.TP
\fBreplace\fR
The \fBreplace\fR mehtod replaces zero or more elements of the list referenced by
\fItobj\fR with the objects in \fIelts\fR ranging from \fIfrom\fR to
\fIto\fR.

.SH "EQUIVALENT C FUNCTIONS"
Tcl_NewListObj, Tcl_ListObjAppendList, Tcl_ListObjLength,
Tcl_ListObjGetElements, Tcl_ListObjIndex, Tcl_ListObjReplace

.SH "SEE ALSO"
InternalRep, TclObject, TclIndex, TclDouble, TclString, TclInteger, TclBoolean, ReflectObject

.SH KEYWORDS
append, index, insert, internal representation, length, list, list object, list type, object, object type, replace


ChangeLog
2019-05-04  Fabrice Ducos <fabrice.ducos@gmail.com>

        * Support of latest versions of Tcl (at least Tcl 8.5
        and Tcl 8.6) and of modern JDK (10+);
        support of JDK 4- abandoned (use of auto-boxing);
        support of OSX systems added;
        details below

        * configuration (tcljava.m4, configure.in),
        and compilation (Makefile.in) templates updated for
        JDK 10+ (javah not available anymore).
        javah is still used when it is available (JDK 9-).
        * minor update of jtclsh.in (for OSX dynamic libraries)
        * many obsolescent constructs (mostly old-style boxing)
        removed in *.java under src/jacl/tcl/lang,
        src/tclblend/tcl/lang and src/tcljava/tcl/lang,
        in order to silence many warnings.
        (effect: support of JDK 4- abandoned)
        * src/native/javaInterp.c: 2010 implementation
        of Java_tcl_lang_Interp_initName() crashed on the
        example given at docs/website/getstart.html,
        because calling Tcl_GetNameOfExecutable
        before Tcl_FindExecutable gives rise to undefined
        behaviour, at least in latest versions of Tcl
        (crashes observed on several systems, at least with
        Tcl 8.6).
        Java_tcl_lang_Interp_initName() is now simplified and
        always calls Tcl_FindExecutable first.
        * all the occurences of interp->result have been
        replaced with Tcl_GetStringResult(interp).
        Accessing interp->result directly was deprecated in
        latest versions of the Tcl C API, and removed from
        Tcl 8.6 (another solution was to #define USE_INTERP_RESULT,
        but this would have been an easy-going, quick and dirty way)
        ref. https://www.tcl.tk/man/tcl/TclLib/Interp.htm

2019-05-01  Fabrice Ducos <fabrice.ducos@gmail.com>

        * tcljava resurrected
        * migration from a CVS to a GIT repository
        (with reposurgeon)

2010-02-26  Mo DeJong  <mdejong@users.sourceforge.net>

	* tests/tcl/regexp.test: Add test case for
        escaped - character in regexp command.
        [Bug 2012389]

2010-02-21  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceAll):
	Fixup regsub -all logic so 5 more tests pass.
	The replaceAll() logic is broken because it depends
	on matching substrings instead of using an index
	into the larger string and Java's regexp library
	uses this to calculate line starts in multiline mode.

2010-02-18  Mo DeJong  <mdejong@users.sourceforge.net>

        Incorporate fix/tests from Tcl head on 2010-02-11,
        aka was Tcl [Bug 2826551, Patch 2948425].

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc):
        Fixup increment index on zero length match issue.
	* tests/tcl/regexp.test: Back port of new regexp tests,
        some tests marked with {endIndex} due to lack of impl.

2010-02-11  Mo DeJong  <mdejong@users.sourceforge.net>

	Patch 2946118

	* src/jacl/tcl/lang/ListCmd.java: Minor style reformat.
	* src/jacl/tcl/lang/Util.java (scanElement, convertElement):
	Remove crusty old code that treated backslash zero
	as the end of the line. That logic was removed from
	C Tcl long ago.
	* tests/inprogress/EventAdaptor.test: Fixup tests
	that depended on backslash zero quirks.
	* tests/tcl/list.test: Add test cases for backslash zero
	as argument to list command.

2009-10-04  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (match, _match,
        replaceFirst, replaceAll): Rewrite -start offset
        logic to correctly deal with odd edge cases.
	* tests/tcl/regexp.test: Add additional test cases
        that cover -start index edge cases.

2009-10-04  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc):
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):
        Rewrite argument processing logic to avoid using
        IndexOutOfBoundsException to deal with an incorrect
        number of arguments.

2009-10-03  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceFirst, replaceAll):
        Fix regsub -all crash because offset was not getting
        advanced by 1 on a match of length zero.
        Fixup -start argument support with and without -all
        in the case where the offset was larger than the
        length of the string.
	* tests/tcl/regexp.test: Add additional test cases.
        Enable regsub test cases that were getting caught
        in an infinite loop.

2009-09-30  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Fixup
        failing tests related to optional subgroups.
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):
        Minor paren fix, no function changes.

2009-09-30  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Rework
        logic related to match result groups.
	* tests/tcl/regexp.test: Add missing test cases for
        non-capturing groups and -inline results.

2009-09-26  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceFirst, replaceAll):
        Fixup use of MULTILINE flag.
	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Fixup
        support for -linestop and -lineanchor arguments to regexp.
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):  Fixup
        support for -linestop and -lineanchor arguments to regsub.
	* tests/tcl/regexp.test: Add missing test cases for
        -linestop and -lineanchor arguments.

2009-09-22  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceFirst, replaceAll):
        Fixup support for the empty string as the string
        argument to regsub by removing the MULTILINE flag when
        an empty string is detected.
	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Fixup
        support for empty string as string argument.
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):
        Remove comments.
	* tests/tcl/regexp.test: Add additional test cases

tcljava/src/jacl/tcl/lang/TclList.java
/*
 * TclList.java
 *
 * Copyright (c) 1997 Sun Microsystems, Inc.
 *
 * See the file "license.terms" for information on usage and
 * redistribution of this file, and for a DISCLAIMER OF ALL
 * WARRANTIES.
 * 
 * RCS: @(#) $Id: TclList.java,v 1.15 2006/06/10 04:15:59 mdejong Exp $
 *
 */

package tcl.lang;

import java.util.ArrayList;

/**
 * This class implements the list object type in Tcl.
 */
public class TclList implements InternalRep {

    /**
     * Internal representation of a list value.
     */
    private ArrayList alist;

    /**
     * Create a new empty Tcl List.
     */
    private TclList() {
	alist = new ArrayList();

	if (TclObject.saveObjRecords) {
	    String key = "TclList";
	    Integer num = (Integer) TclObject.objRecordMap.get(key);
	    if (num == null) {
	        num = 1;
	    } else {
	        num = num + 1;
	    }
	    TclObject.objRecordMap.put(key, num);
	}
    }

    /**
     * Create a new empty Tcl List, with the list pre-allocated to
     * the given size.
     *
     * @param size the number of slots pre-allocated in the alist.
     */
    private TclList(int size) {
	alist = new ArrayList(size);

	if (TclObject.saveObjRecords) {
	    String key = "TclList";
	    Integer num = (Integer) TclObject.objRecordMap.get(key);
	    if (num == null) {
	        num = 1;
	    } else {
	        num = num + 1;
	    }
	    TclObject.objRecordMap.put(key, num);
	}
    }

    /**
     * Called to free any storage for the type's internal rep.
     */
    public void dispose() {
	final int size = alist.size();
	for (int i=0; i<size; i++) {
	    ((TclObject) alist.get(i)).release();
	}
	alist.clear();
    }

    /**
     * DupListInternalRep -> duplicate
     *
     * Returns a dupilcate of the current object.
     *
     */
    public InternalRep duplicate() {
	int size = alist.size();
	TclList newList = new TclList(size);

	for (int i=0; i<size; i++) {
	    TclObject tobj = (TclObject) alist.get(i);
	    tobj.preserve();
	    newList.alist.add(tobj);
	}

	if (TclObject.saveObjRecords) {
	    String key = "TclList.duplicate()";
	    Integer num = (Integer) TclObject.objRecordMap.get(key);
	    if (num == null) {
	        num = 1;
	    } else {
	        num = num + 1;
	    }
	    TclObject.objRecordMap.put(key, num);
	}

	return newList;
    }

    /**
     * Called to query the string representation of the Tcl object. This
     * method is called only by TclObject.toString() when
     * TclObject.stringRep is null.
     *
     * @return the string representation of the Tcl object.
     */
    public String toString() {
	final int size = alist.size();
	if (size == 0) {
	    return "";
	}
	int est = size * 4;

	StringBuffer sbuf = new StringBuffer((est > 64) ? est : 64);
	try {
	    for (int i=0; i<size; i++) {
		Object elm = alist.get(i);
		if (elm != null) {
		    Util.appendElement(null, sbuf, elm.toString());
		} else {
		    Util.appendElement(null, sbuf, "");
		}
	    }
	} catch (TclException e) {
	    throw new TclRuntimeError("unexpected TclException: " + e);
	}

	return sbuf.toString();
    }

    /**
     * Creates a new instance of a TclObject with a TclList internal
     * rep.
     *
     * @return the TclObject with the given list value.
     */

    public static TclObject newInstance() {
	return new TclObject(new TclList());
    }

    /**
     * Called to convert the other object's internal rep to list.
     *
     * @param interp current interpreter.
     * @param tobj the TclObject to convert to use the List internal rep.
     * @exception TclException if the object doesn't contain a valid list.
     */
    private
    static void setListFromAny(Interp interp, TclObject tobj)
	    throws TclException {
	TclList tlist = new TclList();
	splitList(interp, tlist.alist, tobj.toString());
	tobj.setInternalRep(tlist);

	if (TclObject.saveObjRecords) {
	    String key = "TclString -> TclList";
	    Integer num = (Integer) TclObject.objRecordMap.get(key);
	    if (num == null) {
	        num = 1;
	    } else {
	        num = num + 1;
	    }
	    TclObject.objRecordMap.put(key, num);
	}
    }

    /**
     * Splits a list (in string rep) up into its constituent fields.
     *
     * @param interp current interpreter.
     * @param alist store the list elements in this ArraryList.
     * @param s the string to convert into a list.
     * @exception TclException if the object doesn't contain a valid list.
     */
    private static final void splitList(Interp interp, ArrayList alist, String s)
	    throws TclException {
	int len = s.length();
	int i = 0;
	FindElemResult res = new FindElemResult();

	while (i < len) {
	    if (!Util.findElement(interp, s, i, len, res)) {
		break;
	    } else {
		TclObject tobj = TclString.newInstance(res.elem);
		tobj.preserve();
		alist.add(tobj);
	    }
	    i = res.elemEnd;
	}
    }


    /**
     * Tcl_ListObjAppendElement -> TclList.append()
     * 
     * Appends a TclObject element to a list object.
     *
     * @param interp current interpreter.
     * @param tobj the TclObject to append an element to.
     * @param elemObj the element to append to the object.
     * @exception TclException if tobj cannot be converted into a list.
     */
    public static final void append(Interp interp, TclObject tobj,
	    TclObject elemObj)
	    throws TclException {
	if (tobj.isShared()) {
	    throw new TclRuntimeError("TclList.append() called with shared object");
	}
	if (! tobj.isListType()) {
	    setListFromAny(interp, tobj);
	}
	tobj.invalidateStringRep();

	elemObj.preserve();
	((TclList) tobj.getInternalRep()).alist.add(elemObj);
    }

    /**
     * TclList.append()
     *
     * Appends multiple TclObject elements to a list object.
     *
     * @param interp current interpreter.
     * @param tobj the TclObject to append elements to.
     * @param objv array containing elements to append.
     * @param startIdx index to start appending values from
     * @param endIdx index to stop appending values at
     * @exception TclException if tobj cannot be converted into a list.
     */
    static final void append(Interp interp, TclObject tobj,
	    TclObject[] objv,
            final int startIdx, final int endIdx)
            throws TclException {
	if (tobj.isShared()) {
	    throw new TclRuntimeError("TclList.append() called with shared object");
	}
	if (! tobj.isListType()) {
	    setListFromAny(interp, tobj);
	}
	tobj.invalidateStringRep();

	ArrayList alist = ((TclList) tobj.getInternalRep()).alist;

	for (int i = startIdx ; i < endIdx ; i++) {
	    TclObject elemObj = objv[i];
	    elemObj.preserve();
	    alist.add(elemObj);
	}
    }

    /**
     * Queries the length of the list. If tobj is not a list object,
     * an attempt will be made to convert it to a list.
     *
     * @param interp current interpreter.
     * @param tobj the TclObject to use as a list.
     * @return the length of the list.
     * @exception TclException if tobj is not a valid list.
     */
    public static final int getLength(Interp interp, TclObject tobj)
	    throws TclException {
	if (! tobj.isListType()) {
	    setListFromAny(interp, tobj);
	}

	TclList tlist = (TclList) tobj.getInternalRep();
	return tlist.alist.size();
    }

    /**
     * Returns a TclObject array of the elements in a list object.  If
     * tobj is not a list object, an attempt will be made to convert
     * it to a list. <p>
     *
     * The objects referenced by the returned array should be treated
     * as readonly and their ref counts are _not_ incremented; the
     * caller must do that if it holds on to a reference.
     *
     * @param interp the current interpreter.
     * @param tobj the list to sort.
     * @return a TclObject array of the elements in a list object.
     * @exception TclException if tobj is not a valid list.
     */
    public static TclObject[] getElements(Interp interp, TclObject tobj)
	    throws TclException {
	if (! tobj.isListType()) {
	    setListFromAny(interp, tobj);
	}
	TclList tlist = (TclList) tobj.getInternalRep();

	int size = tlist.alist.size();
	TclObject objArray[] = new TclObject[size];
	for (int i=0; i<size; i++) {
	    objArray[i] = (TclObject) tlist.alist.get(i);
	}
	return objArray;
    }

    /**
     * This procedure returns a pointer to the index'th object from
     * the list referenced by tobj. The first element has index
     * 0. If index is negative or greater than or equal to the number
     * of elements in the list, a null is returned. If tobj is not a
     * list object, an attempt will be made to convert it to a list.
     *
     * @param interp current interpreter.
     * @param tobj the TclObject to use as a list.
     * @param index the index of the requested element.
     * @return the the requested element.
     * @exception TclException if tobj is not a valid list.
     */
    public static final TclObject index(Interp interp, TclObject tobj,
	    int index) throws TclException {
	if (! tobj.isListType()) {
	    setListFromAny(interp, tobj);
	}

	TclList tlist = (TclList) tobj.getInternalRep();
	if (index < 0 || index >= tlist.alist.size()) {
	    return null;
	} else {
	    return (TclObject) tlist.alist.get(index);
	}
    }

    /**
     * This procedure inserts the elements in elements[] into the list at
     * the given index. If tobj is not a list object, an attempt will
     * be made to convert it to a list.
     *
     * @param interp current interpreter.
     * @param tobj the TclObject to use as a list.
     * @param index the starting index of the insertion operation. <=0 means
     *   the beginning of the list. >= TclList.getLength(tobj) means
     *   the end of the list.
     * @param elements the element(s) to insert.
     * @param from insert elements starting from elements[from] (inclusive)
     * @param to insert elements up to elements[to] (inclusive)
     * @exception TclException if tobj is not a valid list.
     */
    static final void insert(Interp interp, TclObject tobj, int index,
	    TclObject elements[], int from, int to)
	    throws TclException {
	if (tobj.isShared()) {
	    throw new TclRuntimeError("TclList.insert() called with shared object");
	}
	replace(interp, tobj, index, 0, elements, from, to);
    }

    /**
     * This procedure replaces zero or more elements of the list
     * referenced by tobj with the objects from an TclObject array.
     * If tobj is not a list object, an attempt will be made to
     * convert it to a list.
     *
     * @param interp current interpreter.
     * @param tobj the TclObject to use as a list.
     * @param index the starting index of the replace operation. <=0 means
     *   the beginning of the list. >= TclList.getLength(tobj) means
     *   the end of the list.
     * @param count the number of elements to delete from the list. <=0 means
     *   no elements should be deleted and the operation is equivalent to
     *   an insertion operation.
     * @param elements the element(s) to insert.
     * @param from insert elements starting from elements[from] (inclusive)
     * @param to insert elements up to elements[to] (inclusive)
     * @exception TclException if tobj is not a valid list.
     */
    public static final void replace(Interp interp, TclObject tobj, int index,
	    int count, TclObject elements[], int from, int to)
	    throws TclException {
	if (tobj.isShared()) {
	    throw new TclRuntimeError("TclList.replace() called with shared object");
	}
	if (! tobj.isListType()) {
	    setListFromAny(interp, tobj);
	}
	tobj.invalidateStringRep();
	TclList tlist = (TclList) tobj.getInternalRep();

	int size = tlist.alist.size();
	int i;

	if (index >= size) {
	    // Append to the end of the list. There is no need for deleting
	    // elements.
	    index = size;
	} else {
	    if (index < 0) {
		index = 0;
	    }
	    if (count > size - index) {
		count = size - index;
	    }
	    for (i=0; i<count; i++) {
		TclObject obj = (TclObject) tlist.alist.get(index);
		obj.release();
		tlist.alist.remove(index);
	    }
	}
	for (i=from; i<=to; i++) {
	    elements[i].preserve();
	    tlist.alist.add(index++, elements[i]);
	}
    }

    /**
     * Sorts the list according to the sort mode and (optional) sort command.
     * If tobj is not a list object, an attempt will be made to
     * convert it to a list.
     *
     * @param interp the current interpreter.
     * @param tobj the list to sort.
     * @param sortMode the sorting mode.
     * @param sortIncreasing true if to sort the elements in increasing order.
     * @param command the command to compute the order of two elements.
     * @exception TclException if tobj is not a valid list.
     */

    static void sort(Interp interp, TclObject tobj, int sortMode,
	    int sortIndex, boolean sortIncreasing, String command)
	    throws TclException {
	if (! tobj.isListType()) {
	    setListFromAny(interp, tobj);
	}
	tobj.invalidateStringRep();
	TclList tlist = (TclList) tobj.getInternalRep();

	int size = tlist.alist.size();

	if (size <= 1) {
	    return;
	}

	TclObject objArray[] = new TclObject[size];
	for (int i=0; i<size; i++) {
	    objArray[i] = (TclObject) tlist.alist.get(i);
	}

	QSort s = new QSort();
	s.sort(interp, objArray, sortMode, sortIndex, sortIncreasing, command);

	for (int i=0; i<size; i++) {
	    tlist.alist.set(i, objArray[i]);
	    objArray[i] = null;
	}
    }
}
